openmeter/notification/adapter/channel.go

```diff
@@ -14,167 +14,177 @@ import (
+					NamespacedID: models.NamespacedID{
+						Namespace: params.Namespace,
+						ID:        params.ID,
+					},
+				}
 			}
+
+			return fmt.Errorf("failed to delete notification channel: %w", err)
 		}
 
-		return fmt.Errorf("failed to delete notification channel: %w", err)
+		return nil
 	}
 
-	return nil
+	return entutils.TransactingRepoWithNoValue(ctx, a, fn)
 }
 
-func (r repository) GetChannel(ctx context.Context, params notification.GetChannelInput) (*notification.Channel, error) {
-	db := r.client()
-
-	query := db.NotificationChannel.Query().
-		Where(channeldb.ID(params.ID)).
-		Where(channeldb.Namespace(params.Namespace))
-
-	queryRow, err := query.First(ctx)
-	if err != nil {
-		if entdb.IsNotFound(err) {
-			return nil, notification.NotFoundError{
-				NamespacedID: models.NamespacedID{
-					Namespace: params.Namespace,
-					ID:        params.ID,
-				},
+func (a *adapter) GetChannel(ctx context.Context, params notification.GetChannelInput) (*notification.Channel, error) {
+	fn := func(ctx context.Context, a *adapter) (*notification.Channel, error) {
+		query := a.db.NotificationChannel.Query().
+			Where(channeldb.ID(params.ID)).
+			Where(channeldb.Namespace(params.Namespace))
+
+		queryRow, err := query.First(ctx)
+		if err != nil {
+			if entdb.IsNotFound(err) {
+				return nil, notification.NotFoundError{
+					NamespacedID: models.NamespacedID{
+						Namespace: params.Namespace,
+						ID:        params.ID,
+					},
+				}
 			}
+
+			return nil, fmt.Errorf("failed to fetch notification channel: %w", err)
 		}
 
-		return nil, fmt.Errorf("failed to fetch notification channel: %w", err)
-	}
+		if queryRow == nil {
+			return nil, fmt.Errorf("invalid query result: nil notification channel received")
+		}
 
-	if queryRow == nil {
-		return nil, fmt.Errorf("invalid query result: nil notification channel received")
+		return ChannelFromDBEntity(*queryRow), nil
 	}
 
-	return ChannelFromDBEntity(*queryRow), nil
+	return entutils.TransactingRepo(ctx, a, fn)
 }
 
-func (r repository) UpdateChannel(ctx context.Context, params notification.UpdateChannelInput) (*notification.Channel, error) {
-	db := r.client()
-
-	query := db.NotificationChannel.UpdateOneID(params.ID).
-		SetUpdatedAt(clock.Now().UTC()).
-		SetDisabled(params.Disabled).
-		SetConfig(params.Config).
-		SetName(params.Name)
-
-	queryRow, err := query.Save(ctx)
-	if err != nil {
-		if entdb.IsNotFound(err) {
-			return nil, notification.NotFoundError{
-				NamespacedID: models.NamespacedID{
-					Namespace: params.Namespace,
-					ID:        params.ID,
-				},
+func (a *adapter) UpdateChannel(ctx context.Context, params notification.UpdateChannelInput) (*notification.Channel, error) {
+	fn := func(ctx context.Context, a *adapter) (*notification.Channel, error) {
+		query := a.db.NotificationChannel.UpdateOneID(params.ID).
+			SetUpdatedAt(clock.Now().UTC()).
+			SetDisabled(params.Disabled).
+			SetConfig(params.Config).
+			SetName(params.Name)
```

Comment:
maybe namespace filter? Also if the row is deleted it might be a problem, so at least I would set DeltedAt to nil.